### PR TITLE
Add transform and when functions to actions

### LINKED ISSUE #154: A way to perform final code transformations
I've been adding Plop to my projects over the last few months, and something I've been wanting is an easy way to transform the template output before writing the file. Oftentimes it would be easier/faster to drop some code into the action, rather than abandon the default actions to write a Custom Action Function or a new action. 

Use cases include linting and formatting, or making a one-off change instead of creating and loading a new Handlebars helper.

I found one issue that might relate to this: https://github.com/plopjs/plop/issues/80

One way to accomplish this would be to add an optional `transform()` function to the actions `add`, `addMany`, and `modify`. It would take the template output and the Inquirer data, and should return or resolve the final file contents as a string.

```typescript
function transform(templateOutput: string, data: any): string | Promise<string>
```

A quick look suggests it would be trivial to implement.

https://github.com/plopjs/node-plop/blob/7b53773310d9d9381210bc827a194c2b77e5e820/src/actions/_common-action-add-file.js#L36-L42

I've got some OSS time this week, so if there's no objections I can put together a pull request.

### PR DESCRIPTION
**Pull request by jacksteamdev**
Closes #154.

This adds an optional `transform` function for `add`, `addMany`, and `modify` that allows the action output to be transformed before writing the file to disk.

The `modify` action now can accept one or both of `pattern` and `transform`, but fails with a useful error if neither are provided.

This also adds an optional `when` function for all actions. It was discussed favorably on the main repo (https://github.com/plopjs/plop/issues/101#issuecomment-370511278), but the issue was closed. 

- `when` takes the data object as an argument and skips the action if the function returns a string, which is the reason for skipping the action. A skip is not considered a failure.
- `when` is implemented in the generator runner and runs before the type of action matters, so I've only implemented tests for the `add` action.

`index.d.ts` has been updated. Here's a [usage example gist](https://gist.github.com/jacksteamdev/37175dd3fd80cce8553a025d8d656961).

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -181,6 +181,19 @@ export interface ActionConfig<TData extends object = object> {
...
+	 *
+	 * May also return a Promise which resolves to a string.
+	 *
+	 * The action will continue if action.when()
+	 * returns or resolves to anything other than a string,
+	 * and the return value will be ignored.
+	 *
+	 * @default () => true
+	 */
+	when?: (data: TData) => void | string | Promise<void | string>;
```
Comment (crutchcorn) [inline index.d.ts]: This is super super minor, but I'm noticing that the spacing here isn't consistent with the rest of the `.d.ts` file. Are we able to make that more consistent?

File: src/actions/_common-action-utils.js [hunk 2]
```diff
@@ -49,3 +49,21 @@ export const throwStringifiedError = err => {
...
 	}
 };
+
+export async function getTransformedTemplate(template, data, cfg) {
+	// transform was already typechecked in interface check
+	if ('transform' in cfg) {
+		const result = await cfg.transform(template, data);
+
+		if (typeof result !== 'string')
+			throw new TypeError(
```
Comment (crutchcorn) [inline src/actions/_common-action-utils.js]: I love the check here and the specific error!

File: src/actions/_common-action-utils.js [hunk 3]
```diff
@@ -49,3 +49,21 @@ export const throwStringifiedError = err => {
 		throw err.message || JSON.stringify(err);
 	}
 };
+
+export async function getTransformedTemplate(template, data, cfg) {
+	// transform was already typechecked in interface check
```
Comment (crutchcorn) [inline src/actions/_common-action-utils.js]: We might want to update this comment to be clear that this type-checking is done at the runtime level

File: src/generator-runner.js [hunk 4]
```diff
@@ -114,6 +114,15 @@ export default function (plopfileApi, flags) {
 		// data can also be a function that returns a data object
 		if (typeof cfgData === 'function') { cfgData = await cfgData(); }
 
+		// check if action should run
+		if (typeof cfg.when === 'function') {
+			const stubTrue = () => true;
+			const reasonToSkip = await (cfg.when || stubTrue)(data);
```
Comment (crutchcorn) [inline src/generator-runner.js]: When would `cfg.when` be falsy here? It seems that it's impossible to both have `typeof === 'function'` and it be falsey. Maybe I'm missing something?

File: tests/add-action-when-function.ava.js [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+/**
+ * We are creating files in each test,
+ * but Ava runs tests in parallel by default,
+ * which can make the tests unreliable,
+ * so we need to implement the these tests with test.serial.
+ *
+ * Also we need to clean before each test, b/c we are creating files in each test.
+ * (AvaTest only cleans once before and after the test file)
+ */
```
Comment (crutchcorn) [inline tests/add-action-when-function.ava.js]: Thank you for leaving this here to explain. 💯

Comment (jacksteamdev) [reply tests/add-action-when-function.ava.js]: I should probably copy that into the other test files that need to run serially.

File: index.d.ts (hunk 1 above)
Comment (jacksteamdev) [reply index.d.ts]: Been thinking about formatting, should we add Prettier to a precommit hook?

File: src/generator-runner.js (hunk 4 above)
Comment (jacksteamdev) [reply src/generator-runner.js]: Good catch. That was left over from before it was in an if statement.

**Comment by jacksteamdev** [comment]
@crutchcorn I'll be able to get to the docs on Sunday, but I made the changes you requested. Have a good weekend!

**Comment by crutchcorn** [comment]
Sounds good. You too!

**Comment by jacksteamdev** [comment]
@crutchcorn I've been thinking about changing `when` to `skip`, since the return value is truthy to skip, so don't merge this until I can push that change.

**Comment by jacksteamdev** [comment]
@crutchcorn OK, I've added documentation on the plop repo and renamed the `when` function to `skip`. I think it's ready to go.

File: index.d.ts [hunk 6]
```diff
@@ -181,6 +181,19 @@ export interface ActionConfig<TData extends object = object> {
...
+   *
+   * May also return a Promise which resolves to a string.
+   *
+   * The action will continue if action.when()
+   * returns or resolves to anything other than a string,
+   * and the return value will be ignored.
+   *
+   * @default () => true
+   */
+  when?: (data: TData) => void | string | Promise<void | string>;
```
Comment (amwmedia) [inline index.d.ts]: does this need to be renamed to `skip` as well?

Comment (jacksteamdev) [reply index.d.ts]: Definitely. Thanks for looking!

Comment (jacksteamdev) [reply index.d.ts]: Haha, I forgot to update the types file! This is as good an argument for a full TS conversion as any.

**Pull request merged**